File: app/src/main/java/com/nosfistis/mike/refene/MainActivity.java
package com.nosfistis.mike.refene;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.ActionMode;
import android.view.GestureDetector;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;

import com.nosfistis.mike.refene.database.Refene;
import com.nosfistis.mike.refene.refene.RefenesActivity;
import com.nosfistis.mike.refene.shared.RecyclerViewAdapter;
import com.nosfistis.mike.refene.viewmodel.RefeneViewModel;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class MainActivity extends AppCompatActivity implements ActionMode.Callback {

    private static final int NEW_REFENES = 1;

    private RecyclerView recyclerView;
    private RecyclerViewAdapter viewAdapter;
    private ActionMode actionMode;
    private List<Refene> refeneList;
    private GestureDetector gestureDetector;
    private RefeneViewModel refeneViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        refeneViewModel = ViewModelProviders.of(this).get(RefeneViewModel.class);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(view -> {
            LiveData<Long> insertLiveData = refeneViewModel.getLastInsertId();
            insertLiveData.observe(this, lastInsertedId -> {
                Intent intent = new Intent(this, RefenesActivity.class);
                intent.putExtra(RefenesActivity.REFENE_ID, lastInsertedId);
                startActivityForResult(intent, NEW_REFENES);
                insertLiveData.removeObservers(this);
            });
            refeneViewModel.insert(new Refene());
        });

        recyclerView = findViewById(R.id.my_recycler_view);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addItemDecoration(new DividerItemDecoration(recyclerView.getContext(), layoutManager.getOrientation()));

        gestureDetector = new GestureDetector(this, new RecyclerViewOnGestureListener());
        viewAdapter = new RecyclerViewAdapter(new ArrayList<>(), gestureDetector);
        recyclerView.setAdapter(viewAdapter);
        recyclerView.addOnItemTouchListener(viewAdapter);

        refeneViewModel = ViewModelProviders.of(this).get(RefeneViewModel.class);

        refeneViewModel.getAllRefenes().observe(this, refenes -> {
            refeneList = refenes;

            List<String> refenesNameList = new ArrayList<>();
            for (Refene refene : refeneList) {
                refenesNameList.add(String.valueOf(refene.getId()));
            }

            recyclerView.swapAdapter(new RecyclerViewAdapter(refenesNameList, gestureDetector), false);
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateActionMode(ActionMode actionMode, Menu menu) {
        // Inflate a menu resource providing context menu items
        MenuInflater inflater = actionMode.getMenuInflater();
        inflater.inflate(R.menu.context_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode actionMode, Menu menu) {
        return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode actionMode, MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            case R.id.context_menu_edit:
                actionMode.finish();
                return true;
            case R.id.context_menu_delete:
//                db.open();
//                List<Integer> selections = viewAdapter.getSelectedItems();
//                for (int i : selections) {
//                    db.deleteRefene(refeneList.get(i).getId());
//                    viewAdapter.removeData(i);
//                    //TODO: update refenesList
//                }
//                db.close();
                actionMode.finish();
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onDestroyActionMode(ActionMode actionMode) {
        this.actionMode = null;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == NEW_REFENES && resultCode == RESULT_OK) {
            List<String> refenesNameList = new ArrayList<>();
            for (Refene refene : refeneList) {
                refenesNameList.add(String.valueOf(refene.getId()));
            }

            viewAdapter = new RecyclerViewAdapter(refenesNameList, gestureDetector);
            recyclerView.swapAdapter(viewAdapter, false);
        }
    }

    private class RecyclerViewOnGestureListener extends GestureDetector.SimpleOnGestureListener {
        @Override
        public boolean onSingleTapConfirmed(MotionEvent e) {
            View view = recyclerView.findChildViewUnder(e.getX(), e.getY());
            if (view != null) {
                if (actionMode != null) {
                    viewAdapter.toggleSelection(recyclerView.getChildAdapterPosition(view));
                    return super.onSingleTapConfirmed(e);
                }
                Intent intent = new Intent(view.getContext(), RefenesActivity.class);
                intent.putExtra(RefenesActivity.REFENE_ID, refeneList.get(recyclerView.getChildAdapterPosition(view)).getId());
                startActivity(intent);
            }
            return super.onSingleTapConfirmed(e);
        }

        public void onLongPress(MotionEvent e) {
            if (actionMode != null) {
                return;
            }

            actionMode = startActionMode(MainActivity.this);

            View view = recyclerView.findChildViewUnder(e.getX(), e.getY());
            viewAdapter.toggleSelection(recyclerView.getChildAdapterPosition(Objects.requireNonNull(view)));

            super.onLongPress(e);
        }
    }

}

File: app/src/main/java/com/nosfistis/refene/MainActivity.kt
package com.nosfistis.refene

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.GestureDetector
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.util.Objects

class MainActivity : AppCompatActivity(), ActionMode.Callback {
    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: RecyclerViewAdapter? = null
    private var actionMode: ActionMode? = null
    private var refenesList: List<Long>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        setSupportActionBar(findViewById(R.id.toolbar))

        val fab: FloatingActionButton = findViewById(R.id.fab)
        fab.setOnClickListener {
            val intent = Intent(this, RefenesActivity::class.java)
            intent.putExtra("refID", -1)
            startActivityForResult(intent, NEW_REFENES)
        }

        mRecyclerView = findViewById(R.id.my_recycler_view)

        db = DatabaseHandler(this)
        db!!.open()
        refenesList = db!!.allRefenes
        db!!.close()

        val layoutManager = LinearLayoutManager(this)
        mRecyclerView!!.setHasFixedSize(true)
        mRecyclerView!!.setLayoutManager(layoutManager)
        mRecyclerView!!.addItemDecoration(
            DividerItemDecoration(
                mRecyclerView!!.context,
                layoutManager.orientation
            )
        )

        val refenesNameList: MutableList<String?> = ArrayList()
        for (refenesId in refenesList!!) {
            refenesNameList.add(refenesId.toString())
        }

        val mGestureDetector = GestureDetector(this, RecyclerViewOnGestureListener())
        mAdapter = RecyclerViewAdapter(refenesNameList, mGestureDetector)
        mRecyclerView!!.setAdapter(mAdapter)
        mRecyclerView!!.addOnItemTouchListener(mAdapter!!)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
//
//        if (id == R.id.action_settings) {
//            return true
//        }

        return super.onOptionsItemSelected(item)
    }

    override fun onCreateActionMode(actionMode: ActionMode, menu: Menu): Boolean {
        // Inflate a menu resource providing context menu items
        actionMode.menuInflater.inflate(R.menu.context_menu, menu)
        return true
    }

    override fun onPrepareActionMode(actionMode: ActionMode, menu: Menu): Boolean {
        return false
    }

    override fun onActionItemClicked(actionMode: ActionMode, menuItem: MenuItem): Boolean {
        Log.v("Main", "Inflate")
        when (menuItem.itemId) {
            R.id.context_menu_edit -> {
                actionMode.finish()
                return true
            }

            R.id.context_menu_delete -> {
                db!!.open()
                val selections: List<Int> = mAdapter!!.getSelectedItems()
                for (i in selections) {
                    db!!.deleteRefene(refenesList!![i])
                    mAdapter!!.removeData(i)
                    //TODO: update refenesList
                }
                db!!.close()
                actionMode.finish()
                return true
            }

            else -> return false
        }
    }

    override fun onDestroyActionMode(actionMode: ActionMode) {
        this.actionMode = null
    }

    @Deprecated("This method has been deprecated in favor of using the Activity Result API\n      which brings increased type safety via an {@link ActivityResultContract} and the prebuilt\n      contracts for common intents available in\n      {@link androidx.activity.result.contract.ActivityResultContracts}, provides hooks for\n      testing, and allow receiving results in separate, testable classes independent from your\n      activity. Use\n      {@link #registerForActivityResult(ActivityResultContract, ActivityResultCallback)}\n      with the appropriate {@link ActivityResultContract} and handling the result in the\n      {@link ActivityResultCallback#onActivityResult(Object) callback}.")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == NEW_REFENES && resultCode == RESULT_OK) {
            refenesList = db!!.allRefenes
        }
    }

    private inner class RecyclerViewOnGestureListener : GestureDetector.SimpleOnGestureListener() {
        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            val view = mRecyclerView!!.findChildViewUnder(e.x, e.y)
            if (view != null) {
                if (actionMode != null) {
                    mAdapter!!.toggleSelection(mRecyclerView!!.getChildAdapterPosition(view))
                    return super.onSingleTapConfirmed(e)
                }
                val intent = Intent(view.context, RefenesActivity::class.java)
                intent.putExtra(
                    "refID",
                    refenesList!![mRecyclerView!!.getChildAdapterPosition(view)]
                )
                startActivity(intent)
            }
            return super.onSingleTapConfirmed(e)
        }

        override fun onLongPress(e: MotionEvent) {
            if (actionMode != null) {
                return
            }

            actionMode = startSupportActionMode(this@MainActivity)

            val view = mRecyclerView!!.findChildViewUnder(e.x, e.y)

            if (view != null) {
                mAdapter!!.toggleSelection(
                    mRecyclerView!!.getChildAdapterPosition(
                        Objects.requireNonNull(
                            view
                        )
                    )
                )
            }

            super.onLongPress(e)
        }
    }

    companion object {
        private const val NEW_REFENES = 1

        private var db: DatabaseHandler? = null
    }
}
